src/site/shared/containers/ContextMenu/index.tsx

```diff
@@ -55,7 +58,47 @@ export const ContextMenu = ({info, paste}: Props) => {
             else if (ev.type === "mousedown")
                 dispatch(CloseContextMenu());
         });
-    }, [input])
+    }, [input]);
+
+    /*Position changes are calculated using the react hook so that the
+     *context menu does not jump around during other update events.
+    */
+    useEffect(() => {
+        if(isOpen){
+            let pos = input?.getMousePos();
+
+            if (!menu.current)
+                throw new Error("ContextMenu failed: menu.current is null");
+
+            const offset = 1;
+            const contextMenuWidth = menu.current.getBoundingClientRect().width;
+            const contextMenuHeight = menu.current.getBoundingClientRect().height;
+
+
+            if (pos.x+ contextMenuWidth > window.innerWidth)
+                pos.x -= contextMenuWidth - offset;
+
+            if (pos.y + contextMenuHeight + HEADER_HEIGHT - CONTEXT_MENU_VERT_OFFSET > window.innerHeight)
+                pos.y -= contextMenuHeight - offset;
+
+            //updates position state
+            setPos({ posX:pos.x, posY:pos.y });
+
+            // Update context menu position on canvas
+            menuPos = camera.getWorldPos(input.getMousePos());
+        }
+    }, [isOpen]);
+
+    useDocEvent("mousedown", (ev) => {
+        if (!menu.current)
+            throw new Error("ContextMenu failed: menu.current is null");
+
+        const contextMenuWidth = menu.current.getBoundingClientRect().width;
+        const contextMenuHeight = menu.current.getBoundingClientRect().height;
+
+        if(ev.clientX < posX || ev.clientX > (posX + contextMenuWidth) ||  ev.clientY < posY || ev.clientY > (posY + contextMenuHeight))
+            dispatch(CloseContextMenu());
```

Comment:
I believe you're checking if the mouse position the user clicked is within the bounds of the menu?
If so, I think a much easier way to do this is just check if `menu.current.contains(ev.target)`